### PR TITLE
ENS libs for delegatecall vouchers

### PR DESCRIPTION
**Pull request by ZzzzHui**

### CODE REVIEW

File: contracts/delegatecall/LibEtherForENS.sol
Comment (ZzzzHui) [inline contracts/delegatecall/LibEtherForENS.sol:17]: This address is the same for the networks that ENS root registry has been deployed [so far](https://docs.ens.domains/learn/deployments). It will likely be the same for other future networks as well. But if that's not the case, then this address should be modify accordingly to deploy this library onto that network.

Comment (guidanoli) [reply contracts/delegatecall/LibEtherForENS.sol:17]: I'd be more comfortable with this address being provided in the constructor. This is, considering we convert this library into a contract. Then, we can provide the correct address on the deployment script depending on the chain.

Comment (guidanoli) [reply contracts/delegatecall/LibEtherForENS.sol:17]: One nice experiment would be to see if the contract can access `immutable` variables such as the ENS registry contract even when it's called via `delegatecall`. My intuition tells me that immutable variables are written to byte code at contract creation time.

File: contracts/delegatecall/LibEtherForENS.sol
Comment (guidanoli) [inline contracts/delegatecall/LibEtherForENS.sol:12]: Maybe we could add `@ensdomains/ens-contracts` as a dependency and import these definitions from there?

File: contracts/delegatecall/LibEtherForENS.sol
Comment (guidanoli) [inline contracts/delegatecall/LibEtherForENS.sol:23]: Please prefer using the following function instead for propagating errors.

https://github.com/cartesi/rollups-contracts/blob/2ab9e2388eed4f9f56b0cd2a45be320d5d3e1001/contracts/library/LibError.sol#L9

File: contracts/delegatecall/LibEtherForENS.sol
Comment (guidanoli) [inline contracts/delegatecall/LibEtherForENS.sol:20]: Why not `receiver` instead of `resolution`?

File: contracts/delegatecall/LibEtherForENS.sol
Comment (guidanoli) [inline contracts/delegatecall/LibEtherForENS.sol:15]: Given that we're allowing a `payload` to be passed, this is more like an arbitrary low-level call to an address through its ENS name. Maybe rename this library to `LibCallToENS`?

File: contracts/delegatecall/LibEtherForENS.sol
Comment (ZzzzHui) [reply contracts/delegatecall/LibEtherForENS.sol:20]: Yes we can. It's a different perspective. `recipient` is viewed from the transfer, `resolution` is viewed from the ENS. I will rename it to `recipient`.

File: contracts/delegatecall/LibEtherForENS.sol
Comment (ZzzzHui) [reply contracts/delegatecall/LibEtherForENS.sol:17]: Yeah. `immutable` variables should be in the bytecode during deployment. I have done an experiment to confirm that ✅

File: contracts/delegatecall/CallToENS.sol
Comment (guidanoli) [inline contracts/delegatecall/CallToENS.sol:20]: I'd suggest adding an empty line in-between these two for readability.

File: contracts/delegatecall/CallToENS.sol
Comment (guidanoli) [inline contracts/delegatecall/CallToENS.sol:8]: I'd add an empty line to separate imports from different packages.

File: contracts/delegatecall/CallToENS.sol
Comment (guidanoli) [inline contracts/delegatecall/CallToENS.sol:33]: You can first declare these variables, and then assign them, so that the line fits in 80 characters.

**Review by guidanoli — COMMENTED** [review]
If we decide to take the "all-in-one" approach, I think we could think of a better name for the contract, since we're not really calling addresses by their ENS names in all the cases.

Also, one nice thing I noticed is that this library frees the back-end from providing the application contract address as argument. This information is retrieved at runtime via `address(this)`.

File: contracts/delegatecall/CallToENS.sol
Comment (ZzzzHui) [reply contracts/delegatecall/CallToENS.sol:33]: It's not more than 80 chars in its current way actually

**Comment by ZzzzHui** [comment]
> If we decide to take the "all-in-one" approach, I think we could think of a better name for the contract, since we're not really calling addresses by their ENS names in all the cases.

How about `CallWithENS`?

File: contracts/delegatecall/CallToENS.sol
Comment (guidanoli) [reply contracts/delegatecall/CallToENS.sol:35]: It is 89 chars long, that's why `prettier` is breaking this line.
My suggestion is the following.

```suggestion
        bool success;
        bytes memory returndata;

        (success, returndata) = recipient.call{value: value}(payload);
```

File: test/foundry/dapp/Application.t.sol
Comment (guidanoli) [inline test/foundry/dapp/Application.t.sol:80]: We're missing an array of transfer amounts so that we can actually test the balance after the transfer being non-zero. Also, I think we can come up with better names for these. Why not `_initialSupplies`, `_tokenIds`, and `_transferAmounts`?

Also, I'd be fine with having dynamic state variables only be initialized in `_initVariables`, just like `_interfaceIds`.  We could just loop over the number of token types, and push elements to these arrays one-by-one.

In my opinion, IDs can remain sequential (but then, `_generateUint256Array` wouldn't be necessary, as we'd have the index `i` readily available in the loop anyway, so we could remove that function), total supplies can be constant (and even equal to `_totalSupply`, a sensible default), and transfer amounts should be random and bounded by total supply.

File: test/foundry/dapp/Application.t.sol
Comment (guidanoli) [inline test/foundry/dapp/Application.t.sol:376]: Why not `_testERC721Transfer`?

File: test/foundry/dapp/Application.t.sol
Comment (guidanoli) [inline test/foundry/dapp/Application.t.sol:456]: I don't think we need to separate these into two distinct functions.
We could have just `_testERC20Transfer`, which runs the failure case and then the success case.

File: test/foundry/dapp/Application.t.sol
Comment (guidanoli) [inline test/foundry/dapp/Application.t.sol:379]: I understand you're aiming at minimal diffs for this PR for easier reviewing.
However, before merging, I think we should move the internal functions to the end of the contract.

File: test/foundry/dapp/Application.t.sol
Comment (guidanoli) [inline test/foundry/dapp/Application.t.sol:532]: This function could reuse a good chunk from the `testExecuteEtherTransferVoucher` function. Maybe move the code block from there to an internal function, and call it from here as well?

File: test/foundry/dapp/Application.t.sol
Comment (guidanoli) [inline test/foundry/dapp/Application.t.sol:608]: I feel the lack of regular ERC-1155 transfer tests, without ENS.
Tip: define internal functions to avoid code duplication.

File: test/foundry/util/TestBase.sol
Comment (guidanoli) [inline test/foundry/util/TestBase.sol:45]: We don't need to add this function if we generate all three arrays in the same for-loop.

File: contracts/delegatecall/AssetTransferToENS.sol
Comment (ZzzzHui) [reply contracts/delegatecall/AssetTransferToENS.sol:38]: We could create a library for `safecall`, although I think this code block is already easy to read on its own. Not sure if it improves readability by putting it into a library. Because when reading the codes, readers would need to jump to the library to see what the call is, and then jump back.

File: test/foundry/dapp/Application.t.sol
Comment (ZzzzHui) [reply test/foundry/dapp/Application.t.sol:761]: `keccak256("user.eth")` looks good indeed

File: contracts/delegatecall/AssetTransferToENS.sol
Comment (ZzzzHui) [reply contracts/delegatecall/AssetTransferToENS.sol:1]: If this is a library, then it helps to keep `node` as the first parameter, because then you can do `node.SomeTokenTransfer`. Although now it's not a library, it still acts as a library. And `node` is the "protagonist" of this contract. So I prefer keeping it first actually.

File: test/foundry/dapp/Application.t.sol
Comment (ZzzzHui) [reply test/foundry/dapp/Application.t.sol:456]: It has to be split into 2. Because in the test case `testERC20TransferToENS`, we need to call `_mockENS` between the 2. Because at the end of `_testERC20Fail`, it called `clearMockedCalls`

Comment (ZzzzHui) [reply test/foundry/dapp/Application.t.sol:456]: Although we could call `_mockENS` at the end of `_testERC20Fail` after `clearMockedCalls`, it's not the proper place for `_mockENS` to be in. So I don't mind splitting into 2.

File: test/foundry/dapp/Application.t.sol
Comment (ZzzzHui) [reply test/foundry/dapp/Application.t.sol:718]: Only drawback is then we don't call `vm.clearMockedCalls` at the end of `_mockENS`, but I think it's ok. So I will adapt your suggestion

File: test/foundry/dapp/Application.t.sol
Comment (ZzzzHui) [reply test/foundry/dapp/Application.t.sol:379]: Yes. We should move internal functions to the end, unless it's easier to read at its current place. I think it might be easier to read here, but I don't mind moving it either

File: test/foundry/dapp/Application.t.sol
Comment (ZzzzHui) [reply test/foundry/dapp/Application.t.sol:608]: Agree, let's do that in another issue.

File: contracts/delegatecall/AssetTransferToENS.sol
Comment (ZzzzHui) [reply contracts/delegatecall/AssetTransferToENS.sol:38]: changed my mind :)

Comment (guidanoli) [reply contracts/delegatecall/AssetTransferToENS.sol]: There's some degree of strictness to which we should follow the DRY mantra ("don't repeat yourself"). I think in this case, it helps the user abstract the complexity of low-level operations such as `.call` and `.delegatecall` and handling of return data as error data, which is not very native Solidity-like. Wrapping that in a Solidity library helps abstract implementation details. Also, there's DRY, because this code is repeated here and in `Application`.

File: contracts/delegatecall/AssetTransferToENS.sol
Comment (guidanoli) [reply contracts/delegatecall/AssetTransferToENS.sol:1]: The way I see it, the first parameter should be the contract you'll be calling, which, in this case, is still the token contract. We're just using `node` to resolve it to an Ethereum address, and be able to pass it to a function call to the token contract, therefore playing a secondary role.

File: test/foundry/dapp/Application.t.sol
Comment (guidanoli) [reply test/foundry/dapp/Application.t.sol]: It's just to keep things consistent, though I agree the function would be more distant to where it's being currently used.

File: contracts/delegatecall/AssetTransferToENS.sol
Comment (guidanoli) [reply contracts/delegatecall/AssetTransferToENS.sol:1]: Having said that, in the specific case of `sendCallToENS`, I agree that `node` should be the first parameter, as it will resolve to an address which will be later called, just like a token contract in the other functions.

File: test/foundry/dapp/Application.t.sol
Comment (ZzzzHui) [reply test/foundry/dapp/Application.t.sol]: We don't need to. We used to not call `clearMockedCalls` at all, but started from some PR, we thought it's best practice to close them

File: contracts/delegatecall/AssetTransferToENS.sol
Comment (ZzzzHui) [reply contracts/delegatecall/AssetTransferToENS.sol:1]: I get your perspective :) Our views differ in whether ENS node being a "primary" or "secondary" role. The reason why we have this contract is for various ENS specific cases. Think as if this is a library, and how `node` would be the first parameter if this was a library. I think the same logic still applies even this is not a library. 
Also, the order of the parameters is consistent with the order they are used within the function body too.
Let me know if you think it makes sense or doesn't.

File: test/foundry/dapp/Application.t.sol
Comment (ZzzzHui) [reply test/foundry/dapp/Application.t.sol:464]: Yeah we could do that too

File: test/foundry/dapp/Application.t.sol
Comment (ZzzzHui) [reply test/foundry/dapp/Application.t.sol:713]: I prefer keeping it this way, because:
1. the function doesn't get used elsewhere;
2. by keeping it in the current place, it's easier to see what's the range of the random amount, right?

File: test/foundry/dapp/Application.t.sol
Comment (ZzzzHui) [reply test/foundry/dapp/Application.t.sol:719]: I want to allow 0 `transferAmount` for other tokens. But keeping the first one non-zero helps with the revert case when the app balance is 0, see [here](https://github.com/cartesi/rollups-contracts/blob/43b13e38100a860f528723565d2f0bcfa4402ec3/test/foundry/dapp/Application.t.sol#L635). If we don't make the first transfer amount non-zero,  then it will not be the first token that reverts, it could be the 2nd or 3rd or 4th, which means we need a loop to know which token we expect to revert. Or if all transfer amounts are 0, then no revert will happen.
So I think it's easier to just make the first amount non-zero.

File: test/foundry/dapp/Application.t.sol
Comment (ZzzzHui) [reply test/foundry/dapp/Application.t.sol:646]: Yeah. We can do that too.

File: test/foundry/dapp/Application.t.sol
Comment (guidanoli) [reply test/foundry/dapp/Application.t.sol]: There is no reason to clear them if the test is going to end anyway after that.

File: contracts/delegatecall/AssetTransferToENS.sol
Comment (guidanoli) [reply contracts/delegatecall/AssetTransferToENS.sol:1]: My point is that it shouldn't matter how you identify users, the action is an asset transfer. So, the only difference between an ERC-20 transfer to an Ethereum address and an ERC-20 transfer to an ENS name should be the recipient parameter, whose type changes from an `address` to a `bytes32`. So, from the application's point-of-view, they shouldn't care how this function is implemented, which variable is used first, etc. The application only cares about doing an asset transfer to an ENS name.

For example, let's say we want to transfer 1 CTSI to Alice.
If Alice is identified by her Ethereum address, `0xf39Fd...92266`, then the `DELEGATECALL` voucher would be...

```json
{
    "destination": "<SafeERC20Transfer>",
    "payload": {
        "functionSignature": "safeTransfer(address,address,uint256)",
        "arguments": [
            "<CTSI>",
            "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266",
            1000000000000000000
        ]
    },
}
```

Now, if Alice is identified by her ENS name, `alice.eth`, whose node hash is `0x78719...c3dec`, then I think it would be better if the argument order of the `DELEGATECALL` voucher should be similar.

```json
{
    "destination": "<AssetTransferToENS>",
    "payload": {
        "functionSignature": "sendERC20ToENS(address,bytes32,uint256)",
        "arguments": [
            "<CTSI>",
            "0x787192fc5378cc32aa956ddfdedbf26b24e8d78e40109add0eea2c1a012c3dec",
            1000000000000000000
        ]
    },
}
```

File: test/foundry/dapp/Application.t.sol
Comment (guidanoli) [reply test/foundry/dapp/Application.t.sol:719]: Given that transfer amounts are generated randomly, it is very unlikely that any one of them is zero. But if you want to be extra certain, you can just adjust the minimum transfer amount to be 1 rather than 0, [here](https://github.com/cartesi/rollups-contracts/blob/43b13e38100a860f528723565d2f0bcfa4402ec3/test/foundry/dapp/Application.t.sol#L712-L714).

**Comment by guidanoli** [comment]
This needs a rebase.

File: contracts/delegatecall/AssetTransferToENS.sol
Comment (guidanoli) [inline contracts/delegatecall/AssetTransferToENS.sol:26]: Maybe `sendMessageToENS` or `callENS`?

File: test/foundry/dapp/Application.t.sol
Comment (ZzzzHui) [reply test/foundry/dapp/Application.t.sol]: Yes. I thought the same.

File: test/foundry/dapp/Application.t.sol
Comment (ZzzzHui) [reply test/foundry/dapp/Application.t.sol:719]: Yes, we could do that too.

File: contracts/delegatecall/AssetTransferToENS.sol
Comment (ZzzzHui) [reply contracts/delegatecall/AssetTransferToENS.sol:26]: since it can also be a simple ether transfer, `sendMessageToENS` probably doesn't fit. 
Between `sendCallToENS` and `callENS`, I think I prefer `sendCallToENS ` because all other functions in the contract have names like "sendxxToENS"

**Comment by ZzzzHui** [comment]
rebased. Also moved internal functions to the end.

File: contracts/delegatecall/AssetTransferToENS.sol
Comment (guidanoli) [reply contracts/delegatecall/AssetTransferToENS.sol:26]: I don't know. "Send call" just sounds weird to me, because you wouldn't normally say this in English. You either send something to someone, like a message, or you call them.

> since it can also be a simple ether transfer, sendMessageToENS probably doesn't fit.

I don't understand. Aren't Ether transfers simply message calls with a non-zero amount of Wei?

> because all other functions in the contract have names like "sendxxToENS"

That's one of the reasons why I wanted this function to be in another contract, because it's too general. It's not only suited for Ether transfers, but for arbitrary message calls. If we only care about Ether transfers, we could simply remove the `payload` argument and rename it as `sendEtherToENS`.

Comment (ZzzzHui) [reply contracts/delegatecall/AssetTransferToENS.sol:26]: `sendEtherToENS` should cover most of the cases right? That was the main purpose.
And sending arbitrary calls was just a by-product.

Comment (guidanoli) [reply contracts/delegatecall/AssetTransferToENS.sol:26]: Ok, let's rename it to `sendEtherToENS`. But should we keep the `payload` parameter? It would only be used to call a function of an ENS-identifiable contract, which is very specific, and unlikely, I presume.

Comment (ZzzzHui) [reply contracts/delegatecall/AssetTransferToENS.sol:26]: Yes. Let's keep it. Users can decide what to do with it :)

File: test/foundry/dapp/Application.t.sol
Comment (guidanoli) [inline test/foundry/dapp/Application.t.sol:369]: Shouldn't this be `_testERC20Success`?

File: test/foundry/dapp/Application.t.sol
Comment (guidanoli) [inline test/foundry/dapp/Application.t.sol:377]: Shouldn't this be `_testERC20Fail`?

File: test/foundry/dapp/Application.t.sol
Comment (guidanoli) [inline test/foundry/dapp/Application.t.sol:397]: Did you mean to remove this? The successful case is already tested below.

File: test/foundry/dapp/Application.t.sol
Comment (guidanoli) [inline test/foundry/dapp/Application.t.sol:405]: Did you mean to remove this? This function is already called at setup.

File: test/foundry/dapp/Application.t.sol
Comment (guidanoli) [inline test/foundry/dapp/Application.t.sol:514]: I know that we _know_ that the application contract starts with `_initialSupplies[i]` and the recipient with zero tokens for each token type, but it would be better to construct `appBalances` and `recipientBalances` arrays before the transfer, and use them for the checks after the transfer.

**Review by guidanoli — COMMENTED** [review]
Since we're allowing `sendEtherToENS` to receive a payload, I think it would be good to have a test where the ENS node actually resolves to the address of a contract, and the payload encodes a function call. We could even recycle parte of the "Ether mint" test case for that purpose.

File: test/foundry/dapp/Application.t.sol
Comment (guidanoli) [reply test/foundry/dapp/Application.t.sol]: Fair enough. :-)

File: test/foundry/dapp/Application.t.sol
Comment (ZzzzHui) [reply test/foundry/dapp/Application.t.sol:369]: Yes. Sorry for the silly mistake.

File: .changeset/olive-crabs-carry.md
Comment (guidanoli) [inline .changeset/olive-crabs-carry.md:5]: *for `DELEGATECALL` vouchers
*transfer assets to ENS-identified accounts

**Review by guidanoli — COMMENTED** [review]
I've just noticed we're missing the deployment script of `SafeERC20Transfer` and `AssetTransferToENS`. They could be deployed by a `deploy/03_delegatecall.ts` script.

**Comment by ZzzzHui** [comment]
> I've just noticed we're missing the deployment script of `SafeERC20Transfer` and `AssetTransferToENS`. They could be deployed by a `deploy/03_delegatecall.ts` script.

Yeah. But since `SafeERC20Transfer` is not relevant with this PR, the deploy script will be done in another PR. Ok? :)

**Review by guidanoli — APPROVED** [review]
Great job, LGTM!

**Comment by ZzzzHui** [comment]
Needs @pedroargento 's approval to merge :)

**Review by pedroargento — APPROVED** [review]

**Pull request merged**
